### PR TITLE
Support optional tool method parameters

### PR DESCRIPTION
**Pull request by Kugaaa**
## Issue
Implements #1293 

## Change

This PR adds a new `required` attribute to the `@P` annotation.
Example:

```java
@Tool
public String getWeather(String city, @P(required = false) TemperatureUnit unit) {
    ....
}
```

## General checklist

- [x] There are no breaking changes
- [x] I have added unit and integration tests for my change
- [x] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [x] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

- [x] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

## Checklist for adding new model integration

- [ ] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)

## Checklist for adding new embedding store integration

- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)

## Checklist for changing existing embedding store integration

- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

### CODE REVIEW

**Review by geoand — COMMENTED** [review]
Seems reasonable to me

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/P.java [hunk 1]
```diff
@@ -18,4 +18,10 @@
      * @return the description of a parameter
      */
     String value();
+
+    /**
+     * Whether the parameter is optional
+     * @return true if the parameter is optional, false otherwise
+     */
+    boolean optional() default false;
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/agent/tool/P.java]: Hmm I am wondering if this should be named `required` instead.
Reasoning: in the json schema to which this is eventually translated it is called `required`.

But I am not sure TBH, WDYT?

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/P.java
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/agent/tool/P.java:1]: Please update the documentation (docs/docs/tutorials/6-tools.md) section "### `@P`"

File: langchain4j-core/src/test/java/dev/langchain4j/agent/tool/ToolSpecificationsTest.java
Comment (langchain4j) [inline langchain4j-core/src/test/java/dev/langchain4j/agent/tool/ToolSpecificationsTest.java:1]: It does not seem to test that the parameter is optional/required in the spec

**Review by ghost — COMMENTED** [review]
@Kugaaa thanks a lot!

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/P.java
Comment (Kugaaa) [reply langchain4j-core/src/main/java/dev/langchain4j/agent/tool/P.java:1]: Done, thank you for CR~

File: langchain4j-core/src/test/java/dev/langchain4j/agent/tool/ToolSpecificationsTest.java
Comment (Kugaaa) [reply langchain4j-core/src/test/java/dev/langchain4j/agent/tool/ToolSpecificationsTest.java:1]: I added 2 parameter. `@P(value = "optional", optional = true) int p30`

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/P.java (hunk 1 above)
Comment (Kugaaa) [reply langchain4j-core/src/main/java/dev/langchain4j/agent/tool/P.java]: I changed 'optional' to 'required'.
I defining it as 'optional' was just a habit with the boolean value defaulting to false. But I don't think it's important, like `org.springframework.web.bind.annotation.RequestParam`

File: langchain4j-core/src/test/java/dev/langchain4j/agent/tool/ToolSpecificationsTest.java
Comment (langchain4j) [reply langchain4j-core/src/test/java/dev/langchain4j/agent/tool/ToolSpecificationsTest.java:1]: I misread the change, sorry

**Review by geoand — APPROVED** [review]
LGTM

**Comment by langchain4j** [comment]
@Kugaaa thank you!

**Pull request merged**